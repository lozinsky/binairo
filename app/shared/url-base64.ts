import { expectToBeDefined } from '~/shared/expect';

const PADDING_REGEX = /[.=]{1,2}$/;

const URL_UNSAFE_CHAR_REGEX = /[+/]/g;

const URL_SAFE_CHAR_REGEX = /[-_]/g;

const URL_UNSAFE_CHAR_BY_URL_SAFE_CHAR: Record<string, string> = {
  '-': '+',
  _: '/',
};

const URL_SAFE_CHAR_BY_URL_UNSAFE_CHAR: Record<string, string> = {
  '+': '-',
  '/': '_',
};

export function decodeURLBase64(value: string) {
  return atob(
    value.replaceAll(URL_SAFE_CHAR_REGEX, (substring) =>
      expectToBeDefined(URL_UNSAFE_CHAR_BY_URL_SAFE_CHAR[substring]),
    ),
  );
}

export function encodeURLBase64(value: string) {
  return btoa(value)
    .replaceAll(URL_UNSAFE_CHAR_REGEX, (substring) => expectToBeDefined(URL_SAFE_CHAR_BY_URL_UNSAFE_CHAR[substring]))
    .replace(PADDING_REGEX, '');
}
