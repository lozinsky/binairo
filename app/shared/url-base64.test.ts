import { expect, test } from 'vitest';

import { decodeURLBase64, encodeURLBase64 } from './url-base64';

test.each([
  new Uint8Array([]),
  new Uint8Array([104]),
  new Uint8Array([108, 194, 177]),
  new Uint8Array([89, 44, 194, 170]),
  new Uint8Array([195, 191, 195, 191, 194, 190, 195, 191, 195, 175, 194, 191, 195, 187, 195, 175, 195, 191]),
])('encodes "%s" string to url base64', (value) => {
  expect(encodeURLBase64(new TextDecoder().decode(value))).toMatchSnapshot();
});

test.each(['', 'aA==', 'aA', 'bLE=', 'bLE', 'WSyq', '__--_--_--__'])('decodes "%s" string from url base64', (value) => {
  expect(new TextEncoder().encode(decodeURLBase64(value))).toMatchSnapshot();
});
