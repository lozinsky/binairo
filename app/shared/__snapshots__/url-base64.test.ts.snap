// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`decodes "" string from url base64 1`] = `Uint8Array []`;

exports[`decodes "__--_--_--__" string from url base64 1`] = `
Uint8Array [
  195,
  191,
  195,
  191,
  194,
  190,
  195,
  191,
  195,
  175,
  194,
  191,
  195,
  187,
  195,
  175,
  195,
  191,
]
`;

exports[`decodes "WSyq" string from url base64 1`] = `
Uint8Array [
  89,
  44,
  194,
  170,
]
`;

exports[`decodes "aA" string from url base64 1`] = `
Uint8Array [
  104,
]
`;

exports[`decodes "aA==" string from url base64 1`] = `
Uint8Array [
  104,
]
`;

exports[`decodes "bLE" string from url base64 1`] = `
Uint8Array [
  108,
  194,
  177,
]
`;

exports[`decodes "bLE=" string from url base64 1`] = `
Uint8Array [
  108,
  194,
  177,
]
`;

exports[`encodes "Uint8Array[ 89, 44, 194, 170 ]" string to url base64 1`] = `"WSyq"`;

exports[`encodes "Uint8Array[ 104 ]" string to url base64 1`] = `"aA"`;

exports[`encodes "Uint8Array[ 108, 194, 177 ]" string to url base64 1`] = `"bLE"`;

exports[`encodes "Uint8Array[ 195, 191, 195, 191, 194, 190, 195, 191, 195, 175, 194, 191, 195, 187, 195, 175, 195, 191 ]" string to url base64 1`] = `"__--_--_--__"`;

exports[`encodes "Uint8Array[]" string to url base64 1`] = `""`;
